fastpickle.py

```diff
@@ -488,6 +492,193 @@ def dump(self, obj):
+                                # print("hiii",source[j-1], source[j], j, source[j-10:j+10])
+                                count_of_mark -= 1
+                            defn += source[j]
+                            j += 1
+                        
+                        source = source[:i - 1] + [b'h', binget_code] + source[j:]
+                        break
+
+            # Replacing the binget reference with the extracted definition
+            target = [*map(bytes, zip(target))]
+            target_offset = 0
+            for i, b in enumerate(target):
+                if b == MEMOIZE:
+                    target_offset+=1
+                if b == b'h' and target[i + 1] == binget_code:
+                    target = target[:i] + [*map(bytes, zip(defn))] + target[i + 2:]
+                    break
+            
+            actual_binget_target = 1
+            for i in range(0, target_num):
+                if i not in self.local_memo:
+                    # this happens if the child is not of collection type
+                    continue
+                actual_binget_target += self.local_memo[i]
+            actual_binget_target += target_offset
+            self.old_to_new_binget[binget_code] = actual_binget_target.to_bytes()
+
+            target, source = b''.join(target), b''.join(source)
+            return target, source
+
+        nodes_multiple_accessed = []
+        for id in self.threads_access:
+            if len(set(self.threads_access[id])) > 1:
+                nodes_multiple_accessed.append(id)
+        
+        for id in nodes_multiple_accessed:
+            source_num = self.threads_access[id][0] # thread which contains the defn
+            target_num = min(self.threads_access[id]) # thread which should contain the defn
+            # extract the binget code (after h) in target
+            # source = self.child_bytes[source_num]
+            # target = self.child_bytes[target_num]
+            
+            idx, obj = self.memo[id]
+            _, offset = self.r_memo[idx]
+
+            target_new, source_new = (swap_defn(target_num, source_num, offset, idx, type(obj).__name__))
+
+            self.child_bytes[source_num] = source_new
+            self.child_bytes[target_num] = target_new
+        
+        # # merging
+        print("o to new binget", self.old_to_new_binget)
+
+        # replace with actual binget in every bytecode, except source
+        for i, child in enumerate(self.child_bytes):
+            child_byte_list = [*map(bytes, zip(child))]
+            for idx, b in enumerate(child_byte_list):
+                if b == b'h' and child_byte_list[idx-1] not in int_types:
+                    if child_byte_list[idx+1] not in self.old_to_new_binget:
+                        continue
+                    child_byte_list[idx + 1] = self.old_to_new_binget[child_byte_list[idx+1]]
+            self.child_bytes[i] = b''.join(child_byte_list)
+
+
+        def merge():
```

Comment:
Instead, you can dump the "root" `obj` after dumping all children. That would leave shared memo IDs in the root byte array, which you can swap/replace to fix.

^This is closer to what you'll need to do when we parallelize.